driver-core/src/main/com/mongodb/internal/connection/AuthorizationHeader.java

```diff
@@ -0,0 +1,220 @@
+
+        byte[] kDate = hmac(decodeUTF8("AWS4" + kSecret), decodeUTF8(date));
+        byte[] kRegion = hmac(kDate, decodeUTF8(region));
+        byte[] kService = hmac(kRegion, decodeUTF8(service));
+        byte[] kSigning = hmac(kService, decodeUTF8("aws4_request"));
+
+        return hexEncode(hmac(kSigning, decodeUTF8(toSign)));
+    }
+
+    public String createStringToSign(final String hash, final String timestamp, final String credentialScope) throws SaslException {
+        final List<String> toSign = Arrays.asList(AWS4_HMAC_SHA256, timestamp, credentialScope, hash);
+        return toSign.stream().collect(Collectors.joining("\n"));
+    }
+
+    private String getCredentialScope() throws SaslException {
+        return String.format("%s/%s/%s/aws4_request", getDate(), getRegion(this.host), SERVICE);
+    }
+
+    public String createCanonicalRequest(final byte[] nonce, final String host, final String timestamp) throws SaslException {
+        final String method = "POST";
+        final String uri = "/"; /* unsupported */
+
+        HashMap<String, String> parameters = new HashMap<>();
+        parameters.put("GetCallerIdentity", null);
+        parameters.put("Version", "2011-06-15");
+
+        final String body = getCanonicalQuery(parameters);
+
+        // TODO: add X-MongoDB-Optional-Data if it exist
+        // TODO: add X-Amz-Security-Token if it exist
+        requestHeaders = new HashMap<>();
+        requestHeaders.put("Content-Type", "application/x-www-form-urlencoded");
+        requestHeaders.put("Content-Length", String.valueOf(body.length()));
+        requestHeaders.put("Host", host);
+        requestHeaders.put("X-Amz-Date", timestamp);
+        requestHeaders.put("X-MongoDB-Server-Nonce", Base64.encode(nonce));
+        requestHeaders.put("X-MongoDB-GS2-CB-Flag", "n");
+
+        final String headers = getCanonicalHeaders(requestHeaders);
+        final String signedHeaders = getSignedHeaders(requestHeaders);
+        String hash = hash(body);
+
+        final List<String> request = Arrays.asList(method, uri, "", headers, signedHeaders, hash);
+        return request.stream().collect(Collectors.joining("\n"));
+    }
+
+    public String getSignedHeaders(final Map<String, String> requestHeaders) {
+        return requestHeaders.entrySet().stream()
+                .map(kvp -> kvp.getKey())
+                .map(String::toLowerCase)
+                .sorted()
+                .collect(Collectors.joining(";"));
+    }
+
+    public String getCanonicalHeaders(final Map<String, String> requestHeaders) {
+        return requestHeaders.entrySet().stream()
+                .map(kvp -> String.format("%s:%s", kvp.getKey().toLowerCase(), kvp.getValue()))
+                .sorted()
+                .collect(Collectors.joining("\n", "", "\n"));
+    }
+
+    public String getRegion(final String host) throws SaslException {
+        if (host.matches("(\\w)+") || host.equals("sts.amazonaws.com")) {
```

Comment:
It is faster to use `.indexOf()`, it is simpler and faster than a regex.

Reply:
Done